drivers/sensor/kx022/kx022_trigger.c

```diff
@@ -0,0 +1,315 @@
+ *
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdio.h>
+#include <device.h>
+#include <drivers/i2c.h>
+#include <sys/__assert.h>
+#include <sys/util.h>
+#include <kernel.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+#include "kx022.h"
+
+LOG_MODULE_DECLARE(KX022, CONFIG_SENSOR_LOG_LEVEL);
+
+static void kx022_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct kx022_data *data = CONTAINER_OF(cb, struct kx022_data, gpio_cb);
+	const struct kx022_config *cfg = data->dev->config;
+
+	ARG_UNUSED(pins);
+	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin, GPIO_INT_DISABLE);
+#if defined(CONFIG_KX022_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->trig_sem);
+#elif defined(CONFIG_KX022_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->work);
+#endif
+}
+
+static void kx022_handle_motion_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->motion_handler != NULL) {
+		data->motion_handler(dev, &data->motion_trigger);
+	}
+}
+
+static void kx022_handle_drdy_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->drdy_handler != NULL) {
+		data->drdy_handler(dev, &data->drdy_trigger);
+	}
+}
+static void kx022_handle_slope_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->slope_handler != NULL) {
+		data->slope_handler(dev, &data->slope_trigger);
+	}
+}
+
+static void kx022_handle_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+	const struct kx022_config *cfg = dev->config;
+	uint8_t status, clr;
+
+	if (data->hw_tf->read_reg(data, KX022_REG_INS2, &status) < 0) {
+		return;
+	}
+
+	if (status & 0x02) {
+		printf("===================MOTION DETECTED===============\r\n");
+		kx022_handle_motion_int(dev);
+	} else if (status & 0x10) {
+		printf("======================DATA READY==================\r\n");
+
+		kx022_handle_drdy_int(dev);
+		k_msleep(50);
+
+	} else if (status & 0x01) {
+		printf("==============TILT POSITION DETECTED============\r\n");
+		kx022_handle_slope_int(dev);
+	}
+	data->hw_tf->read_reg(data, KX022_REG_INT_REL, &clr);
+	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin, GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+#ifdef CONFIG_KX022_TRIGGER_OWN_THREAD
+static void kx022_thread(struct kx022_data *data)
+{
+	while (1) {
+		k_sem_take(&data->trig_sem, K_FOREVER);
+		kx022_handle_int(data->dev);
+	}
+}
+#endif
+
+#ifdef CONFIG_KX022_TRIGGER_GLOBAL_THREAD
+static void kx022_work_cb(struct k_work *work)
+{
+	struct kx022_data *data = CONTAINER_OF(work, struct kx022_data, work);
+
+	kx022_handle_int(data->dev);
+}
+#endif
+
+/************************************************************
+ * info: kx022_mode use to set pc1 in standby or operating mode
+ * input: 0 - standby
+ *        1 - operating
+ * **********************************************************/
+
+void kx022_mode(const struct device *dev, bool mode)
```

Comment:
This should be in the main `kx022.c` file. If user chose not to compile the trigger functions this won't be compiled altogether and would result in compile error.